File: 064. Minimum Path Sum/64. Minimum Path Sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int i, j, ret = 0;
        int gridRowSize = grid.size(), gridColSize = grid[0].size();
        int S[gridRowSize][gridColSize];
        if (gridRowSize == 1){
            for (j = 0; j < gridColSize; j++)
                ret += grid[0][j];
            return ret;
        }else if(gridColSize == 1){
            for (i = 0; i < gridRowSize; i++)
                ret += grid[i][0];
            return ret;
        }else{// m > 1 and n > 1
            S[0][0] = grid[0][0];
            for (i = 1; i < gridRowSize; i++)
                S[i][0] = S[i - 1][0] + grid[i][0];
            for (j = 1; j < gridColSize; j++)
                S[0][j] = S[0][j - 1] + grid[0][j];
            for (i = 1; i < gridRowSize; i++)
                for (j = 1; j < gridColSize; j++)
                S[i][j] = grid[i][j] + min(S[i - 1][j],S[i][j - 1]);
            return S[gridRowSize - 1][gridColSize - 1];
        } 
    }
};

File: 064. Minimum Path Sum/64. Minimum Path Sum.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m = len(grid)
        n = len(grid[0])
        if m == 1:
            return sum(grid[0])
        elif n == 1:
            return sum(grid[:][0])
        elif m > 1 and n > 1:
            S = [[0 for c in xrange(n)] for r in xrange(m)]
            S[0][0] = grid[0][0]
            for r in xrange(m):
                S[r][0] = S[r - 1][0] + grid[r][0]
            for c in xrange(n):
                S[0][c] = S[0][c - 1] + grid[0][c]
            for r in xrange(1, m):
                for c in xrange(1, n):
                    S[r][c] = (min(S[r - 1][c], S[r][c - 1]) + grid[r][c])
            return S[r][c]
